<img src=".\images\Logo.png" width=150 align="left" /> <img src=".\images\Logo2.jpg" width=450 align="right" />


# <center><font color= #1e8449 > <b>CU18_Infraestructuras_eventos</font></center>

<font color='steelblue'><b>Citizenlab Data Science Methodology > II - Data Processing Domain </font>
***
> # <font color='steelblue'> <b>05.- Data Collection</font>

<font color='steelblue'>Data Collection is the process to obtain and generate (if required) necessary data to model the problem.</font>

### 05. Agrupar datos de infraestructuras por distrito

* A partir de los datos puntuales, agregar por distrito censal y contar infraestructruras para hacer mapas de regiones. 
* Adicionalmente crear metadatos con descripción y agrupamiento de variables.








<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Settings" data-toc-modified-id="Settings-1"><font color="green">Settings</font></a></span></li><li><span><a href="#Data-Load" data-toc-modified-id="Data-Load-2"><font color="green">Data Load</font></a></span></li><li><span><a href="#ETL-Processes" data-toc-modified-id="ETL-Processes-3">ETL Processes</a></span><ul class="toc-item"><li><span><a href="#Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files" data-toc-modified-id="Import-data-from:-CSV,-Excel,-Tab,-JSON,-SQL,-and-Parquet-files-3.1">Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files</a></span></li></ul></li><li><span><a href="#Synthetic-Data-Generation" data-toc-modified-id="Synthetic-Data-Generation-4">Synthetic Data Generation</a></span></li><li><span><a href="#Fake-Data-Generation" data-toc-modified-id="Fake-Data-Generation-5">Fake Data Generation</a></span></li><li><span><a href="#Open-Data" data-toc-modified-id="Open-Data-6">Open Data</a></span></li><li><span><a href="#Data-Save" data-toc-modified-id="Data-Save-7"><font color="green">Data Save</font></a></span></li><li><span><a href="#Main-Conclusions" data-toc-modified-id="Main-Conclusions-8"><font color=" #2874a6 ">Main Conclusions</font></a></span></li><li><span><a href="#Main-Actions" data-toc-modified-id="Main-Actions-9"><font color=" #2874a6 ">Main Actions</font></a></span><ul class="toc-item"><li><ul class="toc-item"><li><span><a href="#Acciones-done" data-toc-modified-id="Acciones-done-9.0.1"><font color=" #2874a6 ">Acciones done</font></a></span></li><li><span><a href="#Acctions-to-perform" data-toc-modified-id="Acctions-to-perform-9.0.2"><font color=" #2874a6 ">Acctions to perform</font></a></span></li></ul></li></ul></li></ul></div>

## <font color='green'>Settings</font>

### Packages to use

- {tcltk} para selección interactiva de archivos locales
- {readr} para leer y escribir archivos csv
- {dplyr} para explorar datos
- {dityr} para tranformar datos
- {janitor} para limpiar datos


In [24]:
library(readr)
library(dplyr)
library(tidyr)
library(janitor)


### Paths

In [25]:
iPath <- "Data/Input/"
oPath <- "Data/Output/"

## <font color='green'>Data Load</font>

If there are more than one input file, make as many sections as files to import.

Instrucciones
- Los ficheros de entrada del proceso están siempre en Data/Input/.   
- Si hay más de un fichero de entrada, se crean tantos objetos iFile_xx y file_data_xx como ficheros de entrada (xx número correlativo con dos dígitos, rellenar con ceros a la izquierda)

<font color='tomato'><b> OPCION A:</b> Seleccionar fichero en ventana para mayor comodidad</font> 

Data load using the {tcltk} package. Ucomment the line if not using this option

In [26]:
# file_data <- tcltk::tk_choose.files(multi = FALSE)

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>



In [27]:
iFile <- "CU_18_05_04_infraestructuras.csv"
file_data <- paste0(iPath, iFile)

if(file.exists(file_data)){
    cat("Se leerán datos del archivo: ", file_data)
} else{
    warning("Cuidado: el archivo no existe.")
}


Se leerán datos del archivo:  Data/Input/CU_18_05_04_infraestructuras.csv

#### Data file to dataframe

Usar la función adecuada según el formato de entrada (xlsx, csv, json, ...)

In [28]:
data <- read_csv(file_data)


Rows: 1633 Columns: 10
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (8): grupo, tipo, nombre, CODMUN, DIRECCION, info, CMUN, CDIS
dbl (2): X, Y

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


Estructura de  los datos:

In [29]:
glimpse(data)

Rows: 1,633
Columns: 10
$ grupo     <chr> "Transporte", "Transporte", "Transporte", "Transporte", "Tra…
$ tipo      <chr> "Intercambiadores", "Intercambiadores", "Intercambiadores", …
$ nombre    <chr> "Grandes Intercambiadores Plaza Elíptica", "Grandes Intercam…
$ CODMUN    <chr> "079", "079", "079", "079", "079", "079", "079", "079", "079…
$ DIRECCION <chr> "Plaza Elíptica s/n", "Calle Princesa, 89", "Estación Prínci…
$ X         <dbl> -3.716577, -3.719508, -3.719560, -3.689044, -3.676731, -3.68…
$ Y         <dbl> 40.38540, 40.43474, 40.42080, 40.46719, 40.43797, 40.47198, …
$ info      <chr> NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, NA, …
$ CMUN      <chr> "079", "079", "079", "079", "079", "079", "079", "079", "079…
$ CDIS      <chr> "12", "09", "09", "05", "05", "05", "06", "02", "01", "10", …


Muestra de datos:

In [30]:
slice_head(data, n = 5)

grupo      tipo             nombre                                     
1 Transporte Intercambiadores Grandes Intercambiadores Plaza Elíptica    
2 Transporte Intercambiadores Grandes Intercambiadores Moncloa           
3 Transporte Intercambiadores Grandes Intercambiadores Príncipe Pío      
4 Transporte Intercambiadores Grandes Intercambiadores Plaza de Castilla 
5 Transporte Intercambiadores Grandes Intercambiadores Avenida de América
  CODMUN DIRECCION                                X         Y        info CMUN
1 079    Plaza Elíptica s/n                       -3.716577 40.38540 NA   079 
2 079    Calle Princesa, 89                       -3.719508 40.43474 NA   079 
3 079    Estación Príncipe Pío                    -3.719560 40.42080 NA   079 
4 079    Calle San Aquilino y Avenida de Asturias -3.689044 40.46719 NA   079 
5 079    Avenida de América 9 A                   -3.676731 40.43797 NA   079 
  CDIS
1 12  
2 09  
3 09  
4 05  
5 05

## ETL Processes

### Import data from: CSV, Excel, Tab, JSON, SQL, and Parquet files

Se han importado en el apartado Data Load anterior:

* Infraestructuras

Incluir apartados si procede para: Extracción de datos (select, filter), Transformación de datos, (mutate, joins, ...). Si es necesario tratar datos perdidos, indicarlo también en NB 09.2

#### Data Transform

* Contar número de infraestructuras de cada tipo por distrito
* Extender en columnas para caracterizar distritos

In [31]:
tdata_01 <- data |> 
  count(CMUN, CDIS, tipo) |> 
  pivot_wider(names_from = tipo,
              values_from = n, 
              values_fill = 0) 

In [32]:
glimpse(tdata_01)

Rows: 239
Columns: 14
$ CMUN                                      <chr> "001", "002", "003", "004", …
$ CDIS                                      <chr> "01", "01", "01", "01", "01"…
$ `Consultorios de Salud`                   <int> 1, 1, 1, 1, 0, 0, 0, 0, 0, 0…
$ Helisuperficies                           <int> 1, 0, 1, 0, 0, 0, 0, 0, 0, 0…
$ `Centros de Atención a Drogodependientes` <int> 0, 0, 0, 0, 1, 0, 0, 0, 0, 1…
$ `Centros de Salud`                        <int> 0, 0, 0, 0, 2, 3, 2, 1, 2, 5…
$ `Estaciones de Cercanías`                 <int> 0, 0, 0, 0, 1, 0, 1, 1, 0, 2…
$ Hospitales                                <int> 0, 0, 0, 0, 1, 0, 0, 1, 0, 0…
$ `Otros Centros de Salud`                  <int> 0, 0, 0, 0, 1, 0, 0, 0, 0, 0…
$ `Centros de Salud Mental`                 <int> 0, 0, 0, 0, 0, 1, 0, 1, 0, 1…
$ `Centros de Especialidades`               <int> 0, 0, 0, 0, 0, 0, 0, 1, 0, 1…
$ `Bocas de metro`                          <int> 0, 0, 0, 0, 0, 0, 0, 0, 0, 9…
$ Intercambiadores

In [33]:
tdata_01 |> slice_head(n = 5)

CMUN CDIS Consultorios de Salud Helisuperficies
1 001  01   1                     1              
2 002  01   1                     0              
3 003  01   1                     1              
4 004  01   1                     0              
5 005  01   0                     0              
  Centros de Atención a Drogodependientes Centros de Salud
1 0                                       0               
2 0                                       0               
3 0                                       0               
4 0                                       0               
5 1                                       2               
  Estaciones de Cercanías Hospitales Otros Centros de Salud
1 0                       0          0                     
2 0                       0          0                     
3 0                       0          0                     
4 0                       0          0                     
5 1                       1          1                     
  Centros de Salud Mental Centros de Especialidades Bocas de metro
1 0                       0                         0             
2 0                       0                         0             
3 0                       0                         0             
4 0                       0                         0             
5 0                       0                         0             
  Intercambiadores Aeropuertos
1 0                0          
2 0                0          
3 0                0          
4 0                0          
5 0                0

#### Data Extract




* Extraer nombres de columnas


In [34]:
tdata_02 <- data.frame(desc_var = colnames(tdata_01)[3:ncol(tdata_01)])

In [35]:
tdata_02

desc_var                               
1  Consultorios de Salud                  
2  Helisuperficies                        
3  Centros de Atención a Drogodependientes
4  Centros de Salud                       
5  Estaciones de Cercanías                
6  Hospitales                             
7  Otros Centros de Salud                 
8  Centros de Salud Mental                
9  Centros de Especialidades              
10 Bocas de metro                         
11 Intercambiadores                       
12 Aeropuertos

Data Transform

* Limpiar nombres de columnas

In [36]:
tdata_01 <- tdata_01 |> 
  clean_names()

In [37]:
glimpse(tdata_01)

Rows: 239
Columns: 14
$ cmun                                    <chr> "001", "002", "003", "004", "0…
$ cdis                                    <chr> "01", "01", "01", "01", "01", …
$ consultorios_de_salud                   <int> 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, …
$ helisuperficies                         <int> 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, …
$ centros_de_atencion_a_drogodependientes <int> 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, …
$ centros_de_salud                        <int> 0, 0, 0, 0, 2, 3, 2, 1, 2, 5, …
$ estaciones_de_cercanias                 <int> 0, 0, 0, 0, 1, 0, 1, 1, 0, 2, …
$ hospitales                              <int> 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, …
$ otros_centros_de_salud                  <int> 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, …
$ centros_de_salud_mental                 <int> 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, …
$ centros_de_especialidades               <int> 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, …
$ bocas_de_metro                          <int> 0, 0, 0, 0, 0, 0, 0, 0, 0, 9, …
$ intercambiadores

* Completar metadatos

In [44]:
tdata_02 <- tdata_02 |>
    mutate(nombre_var = colnames(tdata_01)[3:ncol(tdata_01)]) |>
    left_join(data |> count(grupo, tipo), 
                by = c("desc_var" = "tipo")) 

In [45]:
glimpse(tdata_02)

Rows: 12
Columns: 4
$ desc_var   <chr> "Consultorios de Salud", "Helisuperficies", "Centros de Ate…
$ nombre_var <chr> "consultorios_de_salud", "helisuperficies", "centros_de_ate…
$ grupo      <chr> "Salud", "Transporte", "Salud", "Salud", "Transporte", "Sal…
$ n          <int> 156, 87, 35, 267, 93, 87, 29, 53, 28, 771, 24, 3


In [46]:
tdata_02

desc_var                               
1  Consultorios de Salud                  
2  Helisuperficies                        
3  Centros de Atención a Drogodependientes
4  Centros de Salud                       
5  Estaciones de Cercanías                
6  Hospitales                             
7  Otros Centros de Salud                 
8  Centros de Salud Mental                
9  Centros de Especialidades              
10 Bocas de metro                         
11 Intercambiadores                       
12 Aeropuertos                            
   nombre_var                              grupo      n  
1  consultorios_de_salud                   Salud      156
2  helisuperficies                         Transporte  87
3  centros_de_atencion_a_drogodependientes Salud       35
4  centros_de_salud                        Salud      267
5  estaciones_de_cercanias                 Transporte  93
6  hospitales                              Salud       87
7  otros_centros_de_salud                  Salud       29
8  centros_de_salud_mental                 Salud       53
9  centros_de_especialidades               Salud       28
10 bocas_de_metro                          Transporte 771
11 intercambiadores                        Transporte  24
12 aeropuertos                             Transporte   3

Si no aplica: Estos datos no requieren tareas de este tipo.

## Synthetic Data Generation

Estos datos no requieren tareas de este tipo.

## Fake Data Generation

Estos datos no requieren tareas de este tipo.

## Open Data

Estos datos no requieren tareas de este tipo.

## <font color='green'>Data Save</font>

Este proceso, puede copiarse y repetirse en aquellas partes del notebbok que necesiten guardar datos.
Recuerde cambiar las cadenas añadida del fichero para diferenciarlas

<font color='tomato'> Identificamos los datos a guardar</font>

1. Infraestructuras agregadas por distrito

In [71]:
data_to_save <- tdata_01


<font color='tomato'>Estructura de nombre de archivos:</font>

* Código del caso de uso, por ejemplo "CU_04"
* Número del proceso que lo genera, por ejemplo "_05".
* Número de la tarea que lo genera, por ejemplo "_01"
* En caso de generarse varios ficheros en la misma tarea, llevarán _01 _02 ... después
* Nombre: identificativo de "properData", por ejemplo "_zonasgeo"
* Extensión del archivo

Ejemplo: "CU_04_05_01_01_zonasgeo.json, primer fichero que se genera en la tarea 01 del proceso 05 (Data Collection) para el caso de uso 04 (vacunas)

Importante mantener los guiones bajos antes de proceso, tarea, archivo y nombre

### Proceso 05

In [72]:
caso <- "CU_18"
proceso <- '_05'
tarea <- "_05"
archivo <- "_01"
proper <- "_infraestructuras_distrito"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [69]:
# file_save_01 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_01 <- paste0(oPath, file_save_01)
# write_csv(data_to_save, path_out_01)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [73]:
file_save_01 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_01 <- paste0(oPath, file_save_01)
write_csv(data_to_save, path_out_01)

cat('File saved as: ')
path_out_01

File saved as: 

[1] "Data/Output/CU_18_05_05_01_infraestructuras_distrito.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [74]:
path_in_01 <- paste0(iPath, file_save_01)
file.copy(path_out_01, path_in_01, overwrite = TRUE)

[1] TRUE

2. Metadatos de infraestructuras

In [77]:
data_to_save_02 <- tdata_02

In [78]:
archivo <- "_02"
proper <- "_infraestructuras_meta"
extension <- ".csv"

<font color='tomato'><b> OPCION A:</b> Uso del paquete "tcltk" para mayor comodidad</font>

* Buscar carpeta, escribir nombre de archivo SIN extensión (se especifica en el código)
* Especificar sufijo2 si es necesario
* Cambiar datos por datos_xx si es necesario

In [58]:
# file_save_02 <- paste0(caso, proceso, tarea, tcltk::tkgetSaveFile(), proper, extension) 
# path_out_02 <- paste0(oPath, file_save_02)
# write_csv(data_to_save_02, path_out_02)

# cat('File saved as: ')
# path_out

<font color='tomato'><b> OPCION B:</b> Especificar el nombre de archivo</font>

- Los ficheros de salida del proceso van siempre a Data/Output/.  

In [79]:
file_save_02 <- paste0(caso, proceso, tarea, archivo, proper, extension) 
path_out_02 <- paste0(oPath, file_save_02)
write_csv(data, path_out_02)

cat('File saved as: ')
path_out_02

File saved as: 

[1] "Data/Output/CU_18_05_05_02_infraestructuras_meta.csv"

#### Copia del fichero a Input

Si el archivo se va a usar en otros notebooks, copiar a la carpeta Input

In [80]:
path_in_02 <- paste0(iPath, file_save_02)
file.copy(path_out_02, path_in_02, overwrite = TRUE)

[1] TRUE

## <font color=' #2874a6 '>Main Conclusions</font>

<font color='steelblue'>List and describe the general conclusions of the analysis carried out​.</font>

### Prerequisites

This working code needs the following conditions:

* For using the interactive selection of file, the {tcltk} package must be installed. It is not needed in production.
* The {readr}, {dplyr}, {tidyr} and {janitor} packages must be installed. 
* The data paths `Data/Input` and `Data/Output` must exist (relative to the notebook path)

### Configuration Management

This notebook has been tested with the following versions of R and packages. It cannot be assured that later versions work in the same way:
* R 4.2.2
* tcltk 4.2.2
* tidyr 1.3.0
* dplyr 1.0.10
* janitor 2.1.0
* readr 2.1.3

### Data structures

#### Objeto `tdata_01`

- Tenemos 239 filas, una por distrito, con los recuentos de cada uno de los tipos de infraestructuras (12)



In [85]:
glimpse(tdata_01)

Rows: 239
Columns: 14
$ cmun                                    <chr> "001", "002", "003", "004", "0…
$ cdis                                    <chr> "01", "01", "01", "01", "01", …
$ consultorios_de_salud                   <int> 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, …
$ helisuperficies                         <int> 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, …
$ centros_de_atencion_a_drogodependientes <int> 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, …
$ centros_de_salud                        <int> 0, 0, 0, 0, 2, 3, 2, 1, 2, 5, …
$ estaciones_de_cercanias                 <int> 0, 0, 0, 0, 1, 0, 1, 1, 0, 2, …
$ hospitales                              <int> 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, …
$ otros_centros_de_salud                  <int> 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, …
$ centros_de_salud_mental                 <int> 0, 0, 0, 0, 0, 1, 0, 1, 0, 1, …
$ centros_de_especialidades               <int> 0, 0, 0, 0, 0, 0, 0, 1, 0, 1, …
$ bocas_de_metro                          <int> 0, 0, 0, 0, 0, 0, 0, 0, 0, 9, …
$ intercambiadores

## Objeto `tdata_02`

- Tenemos 12 filas, una por tipo de infraestructura, con sus metadatos



In [86]:
glimpse(tdata_02)

Rows: 12
Columns: 4
$ desc_var   <chr> "Consultorios de Salud", "Helisuperficies", "Centros de Ate…
$ nombre_var <chr> "consultorios_de_salud", "helisuperficies", "centros_de_ate…
$ grupo      <chr> "Salud", "Transporte", "Salud", "Salud", "Transporte", "Sal…
$ n          <int> 156, 87, 35, 267, 93, 87, 29, 53, 28, 771, 24, 3


#### Observaciones generales sobre los datos

* No aplica

### Consideraciones para despliegue en piloto

* Utilizar los metadatos para etiquetar gráficos y otras salidas.

### Consideraciones para despliegue en producción

* No aplica

## <font color=' #2874a6 '>Main Actions</font>

#### <font color=' #2874a6 '>Acciones done</font>

<font color='steelblue'>Indicate the actions that have been carried out in this process</font>

- Se han agrupado las infraesctructuras por distrito
- Se han calculado el número de infraestructuras por tipo en cada distrito

#### <font color=' #2874a6 '>Acctions to perform</font>

<font color='steelblue'>Indicate the actions that must be carried out in subsequent processes</font>

- Se deben unir los datos de indicadores del INE por distrito

## <font color='RED'>CODE TO DEPLOY (PILOT)</font>

A continuación se incluirá el código que deba ser llevado a despliegue para producción, dado que se entiende efectúa operaciones necesarias sobre los datos en la ejecución del prototipo

<font color=' red '><b>Description</b></font>

- No hay nada que desplegar en el piloto, ya que estos datos son estáticos o en todo caso cambian con muy poca frecuencia, altamente improbable durante el proyecto.


<font color=' red '><b>CODE</b></font>

In [ ]:
# incluir código